### PR TITLE
Add module table builder

### PR DESCRIPTION
**Pull request by sormuras**
Run with: `jshell build.jsh`

[generated/modules.md](https://github.com/sormuras/jpms-module-names/blob/builder/generated/modules.md)

- [x] Add unit tests
- [x] Add already listed projects
- [x] Introduce `pom.xml` parser as default source for project information
- [x] Generate CSV, TSV and `module-maven.properties` + [`module-version.properties`](https://sormuras.github.io/blog/2017-12-13-module-info-properties.html) files

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,8 @@
-csv/.gradle/
-csv/build/
-*.class
+# Gradle
```
Comment (jodastephen) [inline .gitignore:1]: This needs merging with the existing gitignore

File: builder/Builder.java [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+    builder
+        .add("org.joda", "joda-beans", "2.0")
+        .project("Joda-Beans")
+        .homepage("http://www.joda.org/joda-beans");
+
+    builder
+        .add("org.joda", "joda-beans")
+        .project("Joda-Beans")
+        .homepage("http://www.joda.org/joda-beans");
```
Comment (jodastephen) [inline builder/Builder.java]: Maybe the URL could be extracted from the pom.xml

File: builder/Builder.java [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+    static boolean isAutomaticModuleNameAttributeAvailable(ModuleReference moduleReference) {
+      try (ModuleReader moduleReader = moduleReference.open()) {
+        String manifestString =
+            moduleReader
+                .read("META-INF/MANIFEST.MF")
+                .map(StandardCharsets.UTF_8::decode)
+                .map(Object::toString)
+                .orElse("");
+        if (manifestString.contains("Automatic-Module-Name")) {
```
Comment (jodastephen) [inline builder/Builder.java:405]: Will need to check for `module-info.class` too, unless I missed it.

File: .gitignore (hunk 1 above)
Comment (sormuras) [reply .gitignore:1]: Logically it supersedes the existing one. Or did I miss some files?

File: builder/Builder.java (hunk 2 above)
Comment (sormuras) [reply builder/Builder.java]: Good idea - as long there is a) an associated `pom.xml` and b) the URI is specified there.

File: .gitignore (hunk 1 above)
Comment (jodastephen) [reply .gitignore:1]: Is the csv folder gone? I think its still there...

File: builder/Builder.java (hunk 3 above)
Comment (sormuras) [reply builder/Builder.java:405]: That check is _a priori_ done by the `ModuleFinder` usage. Only "automatic" modules are required to be stable, i.e. define an `Automatic-Module-Name` attribute.

File: .gitignore (hunk 1 above)
Comment (sormuras) [reply .gitignore:1]: Entries in `.gitignore` that don't start with a `/` are applied to every sub-folder.

File: builder/Builder.java (hunk 2 above)
Comment (sormuras) [reply builder/Builder.java]: I'll add and use that pom-parsing feature when I implement a logical table section builder.

Something like this:

```java
var section = Section.of("org.joda").project("Joda").homepage("..."). // Section.of(<pom.xml>).
builder.add(section)
```

File: builder/Builder.java (hunk 3 above)
Comment (sormuras) [reply builder/Builder.java:405]: I'll add a flag to the table that indicates whether a module is explicit or autotmatic.

**Comment by sormuras** [comment]
Btw. is it okay if I switch to Java 10 `var` syntax?

**Comment by jodastephen** [comment]
Are we going to need it to run on travis? Anything beyond Java 9 not
supported yet.

**Comment by sormuras** [comment]
Using [install-jdk-10](https://sormuras.github.io/blog/2017-12-08-install-jdk-on-travis.html) it is easy to setup -- but before running on a CI environment, we need a "don't look up already mapped artifacts" feature. Otherwise `mvnrepository` auto-blocks that IP for ... how long I don't know. ;)

**Comment by sormuras** [comment]
Current table versios can viewed here: https://github.com/sormuras/jpms-module-names/blob/builder/generated/modules.md

**Comment by sormuras** [comment]
All generated formats can be viewed here: https://github.com/sormuras/jpms-module-names/tree/builder/generated

- Group crawling, logical section support and Java 10 syntax are not happening soon. ;)
- The `csv` directory can now be marked "deprecated" or removed entirely.
- `BuilderTests` are runnable from within IDEA after marking "builder/" as a source directory and adding "JUnit 5" to the referenced libraries -- a "headless" test task will follow

Anything left to do?

**Pull request merged**